refactor: use frappe.bench for path access

blaggacao opened the pull request:
This PR:
 - accesses site paths via (new) canonical `frappe.bench.site{,s}.path`
 - deprecates access to `frappe.local.site{,s}_path`

blaggacao committed:
deprecate: access to frappe.local.site{,s}_path

blaggacao committed:
refactor: use frappe.bench for path access

blaggacao committed:
deprecate: env {SITES_PATH,FRAPPE_BENCH_ROOT} > FRAPPE_{SITES,BENCH}_PATH

blaggacao committed:
deprecate: non-standard and ad-hoc sites path setter to increase consistency

ankush commented on frappe/deprecation_dumpster.py:
```diff
@@ -200,6 +201,50 @@ def deprecation_warning(marked: str, graduation: str, msg: str):
+			lambda: deprecation_warning(
+				"2024-12-06",
+				"v17",
+				"'local.sites_path' will be deprecated: use 'frappe.bench.sites.path instead'",
```
There is no need to do all this. We can just drop this without warning. 

It's sufficiently "internal" to Frappe framework.

blaggacao replied:
Indeed, I thought so. But I'd like to keep it as a stepping store for my own sake during the refactoring stream (this is meant to be only the first PR). I have an idea though: we could set the graduation to v15, this way it provokes hard errors.

Once the refactoring stream is over and no more errors left, we can remove it immediately.

ankush replied:
It introduces a lot more code than previous version, which honestly is simpler than this at present.

ankush replied:
Also `bencher.py` is weird naming.

blaggacao replied:
> It introduces a lot more code than previous version, which honestly is simpler than this at present.

I'm just a bit afraid at breaking something on the way, I'd prefer to keep it as a temporary band-aid, just to be on the safe side and honor the stability requirement. But if you feel strongly about, we can also drop it right away.

> Also bencher.py is weird naming.

`bench.py` whould have been ideal, but I needed the namespace in `frappe.bench`, so I thought the next best option the file which is responsible for bench layout, i.e. the "maker of bench-compatibility", short "bencher". Idk, I'm not particularly attached to it, but prefer it to `_bench.py` which would slgnal something "internal", which it explicitly is not.

ankush replied:
Has anyone else signed on to this change? 

@akhilnarang or @sumitbhanushali ? 

I'd highly advise against self-merging opinionated changes. I'll be resuming work soon-ish. I am glad to provide an opinion wherever needed. I just don't want to see this codebase become a hodge-podge of 2 different styles.

ankush replied:
E.g. What even is this? Environment variables that depend on class name? https://github.com/frappe/frappe/blob/3d71f594d8b1276777eee168dad4c5dec0148264/frappe/bencher.py#L37-L41

blaggacao replied:
> But if you feel strongly about

Or "sufficiently more confident than me", for that matter. :smile:

ankush replied:
It's well documented: https://github.com/frappe/erpnext/wiki/Naming-Guidelines 

It should at least resemble what it does. `bench_resolver` maybe? (I haven't read entire code)

blaggacao replied:
> I'd highly advise against self-merging opinionated changes.

Not gonna happen. :smile: This had been coordinated at large with @akhilnarang on https://github.com/frappe/frappe/pull/28158 over TG I just ended up hitting the merge button myself.

blaggacao replied:
All actionable feedback has been addressed in followups to #28158

blaggacao committed:
refactor: entirely remove frappe.local.site{,s}_path

see: https://github.com/frappe/frappe/pull/28691#discussion_r1874401242

blaggacao committed:
refactor: drop support for random sites reinstantiation with cwd

An inconsistent behaviour in the first place, this decreases code
readability and by experience judgment of ankush (in TG) was ready to
drop support right away

blaggacao committed:
style: smallfix

blaggacao committed:
fix: get internal apps when scoped

blaggacao committed:
fix: backup folder layout; backup relative to site dir instead of bench

the hard coded 2 level will be a bug in certain scenarios, e.g. when symlinked without compliance

blaggacao commented:
@ankush After a really long debugging session uncovering that backups are archived not relative to site, but to bench, this is now green.

Archiving relative to bench has been breaking symlinked setups, among other things, and is also not compatible with the new path configuration through env variables promise in `frappe/bench_interface.py`

I'd recommend that before iterating on https://github.com/frappe/frappe/pull/28699, we get this merged or at least sanctioned enough to do the next small step of the sister PR

ankush commented:
@akhilnarang  / @sumitbhanushali can one you take this over? (review)

ankush commented:
Lets leave major refactors to core maintainers.
